src/jdk.compiler/share/classes/com/sun/tools/javac/code/Flags.java

```diff
@@ -113,272 +113,127 @@ public static String toString(long flags) {
+     * Flag to indicate the given symbol has been deprecated and marked for removal.
      */
-    public static final long NOT_IN_PROFILE = 1L<<45;
+    public static final long DEPRECATED_REMOVAL = 1L<<19;
 
     /**
-     * Flag that indicates that an override error has been detected by Check.
+     * Flag to indicate the API element in question is for a preview API.
      */
-    public static final long BAD_OVERRIDE = 1L<<45;
+    public static final long PREVIEW_API = 1L<<20; //any Symbol kind
 
     /**
-     * Flag that indicates a signature polymorphic method (292).
+     * Flag to indicate the API element in question is for a reflective preview API.
      */
-    public static final long SIGNATURE_POLYMORPHIC = 1L<<46;
+    public static final long PREVIEW_REFLECTIVE = 1L<<21; //any Symbol kind
 
-    /**
-     * Flag that indicates that an inference variable is used in a 'throws' clause.
+    //classfile flags:
+    /** Flag that marks bridge methods.
      */
-    public static final long THROWS = 1L<<47;
+    public static final long BRIDGE          = 1L<<22;
 
-    /**
-     * Flag that marks potentially ambiguous overloads
+    /** Flag that marks formal parameters.
      */
-    public static final long POTENTIALLY_AMBIGUOUS = 1L<<48;
+    public static final long PARAMETER   = 1L<<23;
 
-    /**
-     * Flag that marks a synthetic method body for a lambda expression
+    /** Flag that marks varargs methods.
      */
-    public static final long LAMBDA_METHOD = 1L<<49;
+    public static final long VARARGS   = 1L<<24;
 
     /**
-     * Flag to control recursion in TransTypes
+     * Flag that marks either a default method or an interface containing default methods.
      */
-    public static final long TYPE_TRANSLATED = 1L<<50;
+    public static final long DEFAULT = 1L<<25;
 
     /**
      * Flag to indicate class symbol is for module-info
      */
-    public static final long MODULE = 1L<<51;
+    public static final long MODULE = 1L<<26;
 
-    /**
-     * Flag to indicate the given ModuleSymbol is an automatic module.
-     */
-    public static final long AUTOMATIC_MODULE = 1L<<52; //ModuleSymbols only
-
-    /**
-     * Flag to indicate the given PackageSymbol contains any non-.java and non-.class resources.
-     */
-    public static final long HAS_RESOURCE = 1L<<52; //PackageSymbols only
-
-    /**
-     * Flag to indicate the given ParamSymbol has a user-friendly name filled.
-     */
-    public static final long NAME_FILLED = 1L<<52; //ParamSymbols only
-
-    /**
-     * Flag to indicate the given ModuleSymbol is a system module.
-     */
-    public static final long SYSTEM_MODULE = 1L<<53;
-
-    /**
-     * Flag to indicate the given symbol has a @Deprecated annotation.
+    //used on trees:
+    /** Flag for synthesized default constructors of anonymous classes.
      */
-    public static final long DEPRECATED_ANNOTATION = 1L<<54;
+    public static final int ANONCONSTR   = 1<<27;
```

Comment:
Question: why this, and other flags that are specific to only one symbol kind have not been translated into the enum? Other examples are `ANONCONSTR_BASED`, `GENERATEDCONSTR`, `COMPACT_RECORD_CONSTRUCTOR`, `THROWS` and probably more.

Reply:
Some of these (IIRC e.g. GENERATEDCONSTR, ANNONCONSTR) are also set on trees, and read and written as bitmasks, so it is more difficult to separate them to e.g. methods only. A few others could possibly be move with a little more code shuffling.

Reply:
I see what you mean - as I said, my worry is that, while this patch addresses the paucity of flags in an effective fashion, the resulting code base end up being worse than it was to begin with, as now a client will have to hunt flags in different enums. If the organization was consistent, I'd have no problem with giving each kind of flag its own enum - but given that the organization is more ad-hoc (e.g. driven by which use cases could be morphed into using enums), I'm unsure.

I mean, what's the advantage of the proposed solution vs. organizing the set of long flags in Flags by categories - so that there's no enum, but so that e.g. a variable symbol flag can reuse the same value as some other method symbol flag?

I think it would deliver same saving, w/o the need to change the flag API. Granted, we also not get any protection for when somebody tries to "merge" flags that are not mergeable - but I'm equally not sure as to whether the proposed solution offers a much more principled alternative?

Reply:
> I see what you mean - as I said, my worry is that, while this patch addresses the paucity of flags in an effective fashion, the resulting code base end up being worse than it was to begin with, as now a client will have to hunt flags in different enums. If the organization was consistent, I'd have no problem with giving each kind of flag its own enum - but given that the organization is more ad-hoc (e.g. driven by which use cases could be morphed into using enums), I'm unsure.
> 
> I mean, what's the advantage of the proposed solution vs. organizing the set of long flags in Flags by categories - so that there's no enum, but so that e.g. a variable symbol flag can reuse the same value as some other method symbol flag?
> 
> I think it would deliver same saving, w/o the need to change the flag API. Granted, we also not get any protection for when somebody tries to "merge" flags that are not mergeable - but I'm equally not sure as to whether the proposed solution offers a much more principled alternative?

I don't think this is a more principled alternative, the main goal here was to reuse the flag bits as safely as possible. Reusing the same bits across various Symbols manually is basically what we do now, but, for me, it is often difficult to prove that a bit is used only for (e.g.) VarSymbols and I can reuse it for MethodSymbols. With this patch, most of the (symbol-specific) checks are done statically, some are done at runtime (via casts), but we can't easily set a method-only flags on a VariableSymbol by accident.